src/DurableTask.Netherite/StorageLayer/Faster/AzureBlobs/AzureStorageDevice.cs

```diff
@@ -603,14 +623,58 @@ async Task WriteToBlobAsync(BlobEntry blobEntry, IntPtr sourceAddress, long dest
                 await this.SingleWriterSemaphore.WaitAsync();
             }
 
-            long offset = 0;
-            while (numBytesToWrite > 0)
+            bool IsFaulted = false;
+
+            try
             {
-                var length = Math.Min(numBytesToWrite, MAX_UPLOAD_SIZE);
-                await this.WritePortionToBlobUnsafeAsync(blobEntry, sourceAddress, destinationAddress, offset, length, id).ConfigureAwait(false);
-                numBytesToWrite -= length;
-                offset += length;
+                long offset = 0;
+                while (numBytesToWrite > 0)
+                {
+                    var length = Math.Min(numBytesToWrite, MAX_UPLOAD_SIZE);
+                    await this.WritePortionToBlobUnsafeAsync(blobEntry, sourceAddress, destinationAddress, offset, length, id).ConfigureAwait(false);
+                    numBytesToWrite -= length;
+                    offset += length;
+                }
+            }
+            catch (Exception e)
+            {
+                IsFaulted = true;
+
+                // details about failed storage accesses are already traced by the error handler, but we still create another trace here
+                // to help us debug situations where exceptions are thrown in other places, and to clarify the execution path
+                this.BlobManager.StorageTracer?.FasterStorageError($"{nameof(WriteToBlobAsync)} id={id} encountered exception", e);
             }
+            finally
+            {
+                if (this.underLease)
+                {
+                    this.SingleWriterSemaphore.Release();
+                }
```

Comment:
nit: please add comment on why this is important to do at this point

Reply:
Also, do we not have a forceful semaphore release if it hangs for too long?

Reply:
> lease add comment on why this is important to do at this point
ok

 > do we not have a forceful semaphore release if it hangs for too long?
the hang check is somewhere else. We don't need it here since the entire object (incl the semaphore) gets disposed when the partition is terminated. Also, have not seen Azure Storage hanging recently.